x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/slowlog/EsqlSlowLog.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.slowlog;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.logging.ESLogMessage;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xcontent.json.JsonStringEncoder;
+import org.elasticsearch.xpack.core.security.SecurityContext;
+import org.elasticsearch.xpack.core.security.user.User;
+import org.elasticsearch.xpack.esql.session.Result;
+
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.elasticsearch.xpack.esql.plugin.EsqlPlugin.ESQL_SLOWLOG_THRESHOLD_INCLUDE_USER_SETTING;
+import static org.elasticsearch.xpack.esql.plugin.EsqlPlugin.ESQL_SLOWLOG_THRESHOLD_QUERY_DEBUG_SETTING;
+import static org.elasticsearch.xpack.esql.plugin.EsqlPlugin.ESQL_SLOWLOG_THRESHOLD_QUERY_INFO_SETTING;
+import static org.elasticsearch.xpack.esql.plugin.EsqlPlugin.ESQL_SLOWLOG_THRESHOLD_QUERY_TRACE_SETTING;
+import static org.elasticsearch.xpack.esql.plugin.EsqlPlugin.ESQL_SLOWLOG_THRESHOLD_QUERY_WARN_SETTING;
+
+public final class EsqlSlowLog {
+
+    private final ClusterSettings clusterSettings;
+
+    public static final String SLOWLOG_PREFIX = "esql.slowlog";
+
+    private static final Logger queryLogger = LogManager.getLogger(SLOWLOG_PREFIX + ".query");
+    private final SecurityContext security;
+
+    private volatile long queryWarnThreshold;
+    private volatile long queryInfoThreshold;
+    private volatile long queryDebugThreshold;
+    private volatile long queryTraceThreshold;
+
+    private volatile boolean includeUser;
+
+    public EsqlSlowLog(ClusterSettings settings, SecurityContext security) {
+        settings.initializeAndWatch(ESQL_SLOWLOG_THRESHOLD_QUERY_WARN_SETTING, this::setQueryWarnThreshold);
+        settings.initializeAndWatch(ESQL_SLOWLOG_THRESHOLD_QUERY_INFO_SETTING, this::setQueryInfoThreshold);
+        settings.initializeAndWatch(ESQL_SLOWLOG_THRESHOLD_QUERY_DEBUG_SETTING, this::setQueryDebugThreshold);
+        settings.initializeAndWatch(ESQL_SLOWLOG_THRESHOLD_QUERY_TRACE_SETTING, this::setQueryTraceThreshold);
+        settings.initializeAndWatch(ESQL_SLOWLOG_THRESHOLD_INCLUDE_USER_SETTING, this::setIncludeUser);
+        this.clusterSettings = settings;
+
+        this.security = security;
+    }
+
+    public EsqlSlowLog(ClusterSettings settings) {
+        this(settings, null);
+    }
+
+    public void onQueryPhase(Result esqlResult, String query) {
```

Comment:
This method and onQueryFailure are very similar - same code should be reused across both of them.